app/celery/signals.py

```diff
@@ -0,0 +1,103 @@
+import inspect
+import asyncio
+from celery.signals import (
+    task_received, task_internal_error, task_revoked,
+    task_success, task_failure, task_retry, task_prerun
+)
+from celery_singleton.util import generate_lock
+from celery_singleton.config import Config
+from celery_singleton.backends import get_backend
+from app.celery.worker import celery_app
+from app.database import SessionLocal
+from app.services import WebsocketManagerService, JobStatusService, InstructorService
+from app.models.job_status import JobStatusModel, JobStatus
+from app.schemas import JobSchema, WebsocketJobStatusMessage
+
+def handle_job_status_update(job_id: str, job_type: str | None, job_status: JobStatus) -> None:
+    """ Store a job status update in the database and dispatch a websocket event. """
+    async def _handle_job_status_update():
+        with SessionLocal() as session:
+            job_status_model = JobStatusService(session).create_job_status_update(job_id, job_type, job_status)
+            ws_message = WebsocketJobStatusMessage(
+                data=JobSchema.from_orm(job_status_model)
+            )
+            
+            if job_type == "downsync":
+                user_list = await InstructorService(session).list_instructors()
+            else:
+                user_list = []
+
+            WebsocketManagerService.publish_sync_pubsub_ws_message(ws_message, user_list)
+    
+    asyncio.run(_handle_job_status_update())
+
+@task_received.connect
+def task_received_handler(request, **kw):
+    handle_job_status_update(request.task_id, request.task_name, JobStatus.RECEIVED)
+
+@task_prerun.connect
+def task_prerun_handler(sender=None, task_id=None, **kw):
+    handle_job_status_update(task_id, sender.name, JobStatus.STARTED)
+
+@task_success.connect
+def task_success_handler(sender=None, result=None, **kw):
+    handle_job_status_update(sender.request.id, sender.name, JobStatus.SUCCESS)
+
+@task_failure.connect
+def task_failure_handler(sender=None, task_id=None, **kw):
```

Comment:
There's an argument that can be passed into this function called `exception` that contains the exception that's raised. Will that exception be bubbled up where it will be properly logged, or do we need to log it here?

Reply:
It is bubbled up and logged automatically, so no need to do so here.